diff --git a/Pipeto/src/commands/history.c b/Pipeto/src/commands/history.c
index acc5687..98263d4 100644
--- a/Pipeto/src/commands/history.c
+++ b/Pipeto/src/commands/history.c
@@ -53,12 +53,11 @@ void history_clear(history_t *history)
 
 void history_free(history_t *history)
 {
-    for (int i = 0; i < history->history_count; i++)
+    for (int i = 0; i < history->history_count - 1; i++)
         free(history->array[i]);
-    free(history);
-    history = NULL;
     history->history_count = 0;
     history->history_size = 0;
+    history = NULL;
 }
 
 char *history_get(int index, history_t *history)
