diff --git a/Pipeto/src/commands/unlock_secret_mode.c b/Pipeto/src/commands/unlock_secret_mode.c
index 1739714..4d6ef65 100644
--- a/Pipeto/src/commands/unlock_secret_mode.c
+++ b/Pipeto/src/commands/unlock_secret_mode.c
@@ -12,9 +12,13 @@
 
 void unlock_secret_mode(bool *is_admin, UNUSED history_t *history)
 {
+    char mdp[13] = {0};
+
+    crypt_password(mdp, 12, '{', 'A', 'D', 'M', 'I', 'N', '4',
+    '2', '4', '2', '}', '\0');
     if (*is_admin == true) {
         printf("Secret mode unlocked! Welcome, admin.\n");
-        printf("{ADMIN4242}\n");
+        printf("%s\n", mdp);
     } else {
         printf("Access denied. You do not have the required privileges.\n");
     }
